File: webapp/src/lib/blog/tableBlockLimits.ts
export const MAX_TABLE_ROWS = 20;
export const MAX_TABLE_COLS = 5;
export const MAX_TABLE_CELL_CHARS = 4000;
export function tableHasContent(rows: string[][]): boolean {
  return rows.some((row) => row.some((cell) => cell.trim().length > 0));
}
export function clampTableMatrix(
  rows: string[][],
  maxRows: number,
  maxCols: number,
  maxCellChars: number,
): string[][] {
  if (!rows.length) return [[""]];
  const w = Math.min(maxCols, Math.max(1, ...rows.map((r) => r.length)));
  const h = Math.min(maxRows, Math.max(1, rows.length));
  return rows.slice(0, h).map((r) => {
    const cells = r
      .slice(0, w)
      .map((c) => (c.length > maxCellChars ? c.slice(0, maxCellChars) : c));
    while (cells.length < w) cells.push("");
    return cells;
  });
}
export function truncateCell(value: string, maxCellChars: number): string {
  if (value.length <= maxCellChars) return value;
  return value.slice(0, maxCellChars);
}
export function tableEffectiveColCount(rows: string[][]): number {
  if (!rows.length) return 1;
  let width = Math.max(1, ...rows.map((row) => row.length));
  while (width > 1) {
    const trailingEmpty = rows.every(
      (row) => width - 1 >= row.length || row[width - 1].trim().length === 0,
    );
    if (!trailingEmpty) break;
    width--;
  }
  return width;
}

File: mobile/lib/utils/table_block_limits.dart
const int maxTableRows = 20;
const int maxTableCols = 5;
const int maxTableCellChars = 4000;

List<List<String>> clampTableMatrix(List<List<String>> rows) {
  if (rows.isEmpty) return [['']];

  final width = rows
      .map((r) => r.length)
      .fold<int>(1, (a, b) => a > b ? a : b)
      .clamp(1, maxTableCols);
  final height = rows.length.clamp(1, maxTableRows);

  return List.generate(height, (ri) {
    final src = ri < rows.length ? rows[ri] : <String>[];
    final cells = List.generate(width, (ci) {
      final raw = ci < src.length ? src[ci] : '';
      return raw.length > maxTableCellChars ? raw.substring(0, maxTableCellChars) : raw;
    });
    return cells;
  });
}

bool tableHasContent(List<List<String>> rows) {
  for (final row in rows) {
    for (final cell in row) {
      if (cell.trim().isNotEmpty) return true;
    }
  }
  return false;
}

bool tableWithinLimits(List<List<String>> rows) {
  if (rows.length > maxTableRows) return false;
  for (final row in rows) {
    if (row.length > maxTableCols) return false;
  }
  return true;
}

/// Columns to show in preview — drops trailing columns that are empty in every row.
int tableEffectiveColCount(List<List<String>> rows) {
  if (rows.isEmpty) return 1;
  var width = rows.fold<int>(1, (w, row) => row.length > w ? row.length : w);
  while (width > 1) {
    final trailingEmpty = rows.every(
      (row) => width - 1 >= row.length || row[width - 1].trim().isEmpty,
    );
    if (!trailingEmpty) break;
    width--;
  }
  return width;
}
